### DISCUSSION TITLE
Caps Lock as Escape and Control

### DISCUSSION THREAD

**Discussion by sezanzeb** [sezanzeb/input-remapper#34] [Ideas]
https://github.com/sezanzeb/key-mapper/issues/30

**Comment by todfox** [comment]
I would love to see this. For me it is using the tab key as tab if pressed, super_l when held. 

h(k(super_l)).k(tab).w(5) kinda works, but its hit and miss

**Reply by sezanzeb** [reply]
there is no way to differentiate "pressing" and "holding" purely based on the event values (1 for down, 0 for up). Every key press means that it is also held for some time.

please specify again what you are trying to achieve. The op is talking about "`a` when released without a combination, `b` when combined with any arbitrary key"

**Reply by todfox** [reply]
xkb_symbols "hypers" {
    key  TAB { [ Hyper_L, Hyper_L ] };
    key BKSL { [ Hyper_R, Hyper_R ] };
    key I252 { [ Tab,	ISO_Left_Tab ] };
    key I253 { [ backslash, bar ] };
    modifier_map Mod4 { Super_L, Super_R, Hyper_L, Hyper_R };
};

**Reply by todfox** [reply]
https://github.com/altercation/dotfiles-tilingwm/blob/master/bin/wm/init-keyboard

**Comment by sezanzeb** [comment]
Maybe I can offer plugin support.

People then can just write whatever they like using python and offer their plugins in github, for example to provide the functionality that you desire.

but I need to think about it for some more time. Where to run them, how to load them and which information to provide to them.

And also the security implications. The service is running as root and it shouldn't be able to run plugins that are accessible by non-root users, so offering a plugin directory in home is out of the question.

**Reply by todfox** [reply]
I want to move to Wayland from X but just can't leave behind Xcape. This would make you my hero! As such, let me know now to buy you a couple cups of coffee.

**Comment by guy-who-googles** [comment]
In the most recent version, I tried 2 approaches to this:

1. One Entry, but unreliable
* `Caps_Lock = if_single(key(Escape), modify(Control_L,hold()),200)`
    * single press has to be pressed before timeout (200ms)
    * chorded press has to have duration of timeout (200ms)

1. Many Entries, but reliable
* `Caps_Lock = if_single(key(Escape),)`
* for every key: `Caps_Lock+<key> = modify(Control_L, key(<key>))`
    * i.e. `Caps_Lock+c = modify(Control_L, key(c))`

**Reply by sezanzeb** [reply]
instead of `modify(Control_L,hold())` you can write `hold(Control_L)`

**Reply by sezanzeb** [reply]
in your reliable method you are not using timeouts, you also don't have to use a timeout for `if_single` in the first example (depending on what you are trying to achieve)

**Reply by guy-who-googles** [reply]
Thanks for the tip! As suggested, `Caps_Lock = if_single(key(Escape),hold(Control_L))` does the trick.

I originally chose to put a timeout for `if_single` because the unreliable solution without it took too long. I incorrectly assumed that the default timeout was 1000 ms, based on the docs. The docs are clear though, I was just mistaken.

**Comment by sezanzeb** [comment]
`if_single` should do the trick. See https://github.com/sezanzeb/key-mapper/issues/183 on how to do a "space cadet shift"

**Reply by Bad3r** [reply]
Sorry I am confused. Is it possible  to remap caps lock to behave like escape on press and control on hold in the latest version?

If so is it possible via the GUI?

**Reply by sezanzeb** [reply]
`if_tap(key(Escape), key(Control_L))`
